solutions/YS Vertex Set Path Composite.mdx

```diff
@@ -0,0 +1,251 @@
+---
+id: ys-VertexSetPathComposite
+source: YS
+title: Vertex Set Path Composite
+author: Andrew Cheng
+---
+
+**Time Complexity:** $\mathcal{O}(N \log^2 N)$ from HLD
+
+**Main Idea:**
+Different from a classic HLD problem, the nesting of functions does not follow the commutative property.
+In other words, the answer of the query $1$ $u$ $v$ $x$ is different from the answer of the query $1$ $v$ $u$ $x$.
+
+To solve this problem, we can break down the path from node $u$ to node $v$ into two paths: from node $u$ to $LCA(u,v)$ and from $LCA(u,v)$ to node $v$.
+The path from node $u$ to $LCA(u,v)$ will decrease in depth, and the path from $LCA(u,v)$ to node $v$ will increase in depth.
+
+We can then maintain two segment trees
+
+The first segment tree will calculate the equivalent function that results from moving from node $u$ to $LCA(u,v)$. And the second segment tree
+will calculate the equivalent function that results from moving from $LCA(u,v)$ to node $v$. After this we can plug in the given $x$ into the two
+functions to get the answer of the query.
+
+It should be mentioned that the function at node $LCA(u,v)$ should only be calculated once. To avoid calculating the node twice, we can intentionally avoid $LCA(u,v)$ during
+exactly one of the two path queries.
+
+The update queries can be implemented using segment tree point updates on both segment trees.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+
+#include <bits/stdc++.h>
+#define ll long long
+#define pi pair<ll,ll>
+#define fi first
+#define se second
+#define pb push_back
+using namespace std;
+
+const int maxN = 2e5+5;
+const int height = 25;
+const int MOD = 998244353;
+
+int N,Q;
+vector<int> adj[maxN];
+int dep[maxN];
+int top[maxN];
+int sz[maxN];
+int son[maxN];
+int fa[maxN];
+int dfn[maxN];
+int dp[maxN][height];
+pi ori [maxN];
+pi val [maxN];
+int cnt;
+
+//LCA
+int LCA(int x, int y){
+	if(dep[x] < dep[y])swap(x,y);
+	for(int i = height-1; i >= 0; --i){
+		if(dep[x] - (1<<i) >= dep[y]){
+			x = dp[x][i];
+		}
+	}
+	if(x == y)return x;
+	for(int i = height-1; i >= 0; --i){
+		if(dp[x][i] != dp[y][i]){
+			x = dp[x][i];
+			y = dp[y][i];
+		}
+	}
+	return dp[x][0];
+}
+
+//Segment Tree
+
+#define lc p<<1
+#define rc (p<<1)+1
+
+pi tree[maxN<<2]; //Going up
+pi tree2[maxN<<2]; //Going down
+
+pi merge(pi x, pi y){
+	//f(x) = ax + b; g(x) = cx + d
+	//Then f(g(x)) = a(cx+d) + b = ac*x + b + ad
+	pi k;
+	k.fi = (x.fi * y.fi)%MOD;
+	k.se = ((x.fi * y.se)%MOD + x.se)%MOD;
+	return k;
+}
+
+void pushup(int p){
+	tree[p] = merge(tree[lc],tree[rc]);
+	tree2[p] = merge(tree2[rc],tree2[lc]);
+	return;
```

Comment:
why do you have `return;` at the end of some of these functions

Reply:
just asking is it a good habit to return every function ? Or is it completely unnecessary?

Reply:
I think it's not necessary